common/parallel/TreeCrawler.go

```diff
@@ -90,25 +127,228 @@ func (c *crawler) runWorkersToCompletion(ctx context.Context) {
+		}
 		if err != nil {
 			c.output <- CrawlResult{err: err}
 			// output the error, but we don't necessarily stop the enumeration (e.g. it might be one unreadable dir)
 		}
 	}
 }
 
+func (c *crawler) receiveTqueue(root Directory) {
+
+	for ch := range c.tqueue {
+		c.cond.L.Lock()
+
+		if ch != nil {
+			c.unstartedDirs = append(c.unstartedDirs, ch)
+		}
+		c.cond.L.Unlock()
+		c.cond.Broadcast()
+	}
+
+	c.cond.L.Lock()
+	c.tqueueClosed = true
+	c.cond.L.Unlock()
+	c.cond.Broadcast()
+}
+
+// TODO: Need to pass it parallel crawl.
+const SizeInGB = 1024 * 1024 * 1024
+
+func (c *crawler) autoPacerWait(ctx context.Context) {
+	// Return in case its not sync operation.
+	if !c.isSync {
+		return
+	}
+
+	// Destination case where we return if tqueue channel closed or if some directories need processing.
+	// TODO: It has one drawback if destination is faster than source it will lead to closing of go-routine threads which we don't want.
+	if c.isSync && !c.isSource {
+		for {
+			c.cond.L.Lock()
+			if len(c.unstartedDirs) == 0 && !c.tqueueClosed && ctx.Err() == nil {
+				c.cond.L.Unlock()
+				time.Sleep(100 * time.Millisecond)
+				continue
+			}
+			c.cond.L.Unlock()
+			return
+		}
+	}
+	//
+	// Consider 80% full as low-water-mark. Anything less than 80% full is fair game and we let the source traverser
+	// run full throttle. Once it exceeds low-water-mark we slowly start taking the foot off the pedal and every thread/goroutine
+	// needs to wait for a small time before proceeding. Once it touches/crosses high-water-mark no scanner thread/goroutine will
+	// be allowed to proceed. For reasonable sized directories this should ensure that we remain in MaxObjectIndexerSizeGB limit
+	// for most practical scenarios. For large/huge directories we might exceed MaxObjectIndexerSizeGB by the size of the largest directory.
+	//
+	MaxObjectIndexerSizeGB := int64(c.maxObjectIndexerSizeInGB * SizeInGB)
+	lowWaterMarkMB := (MaxObjectIndexerSizeGB * 8) / 10
+
+	highWaterMarkMB := MaxObjectIndexerSizeGB
+
+	mapSize := c.getIndexerMapSize()
+
+	// Nice sunny morning, press that pedal more.
+	if mapSize < lowWaterMarkMB {
+		return
+	}
+
+	//
+	// We are nearing danger zone, start slowing down.
+	//
+	// TODO: The sleep duration may be decided based on how fast the
+	//       objectIndexerMap is seen to be growing.
+	//
+	if mapSize < highWaterMarkMB && ctx.Err() == nil {
+		time.Sleep(1 * time.Second)
+		return
+	}
+
+	// In danger zone. Don't proceed any further without dropping the objectIndexerMap size below lowWaterMarkMB.
+	for mapSize > lowWaterMarkMB && ctx.Err() == nil {
+		time.Sleep(1 * time.Second)
+	}
+
+	return
+}
+
+func (c *crawler) processOneDirectoryWithAutoPacer(ctx context.Context, workerIndex int) (bool, error) {
+	const maxQueueDirectories = 1000 * 1000
+	const maxQueueDirsForBreadthFirst = 100 * 1000 // figure is somewhat arbitrary.  Want it big, but not huge
+
+	var toExamine Directory
+	stop := false
+
+	c.autoPacerWait(ctx)
```

Comment:
don't call autopacer for target, and add a comment above this

// Before picking a new directory to enumerate, call autoPacerWait() to check if we need to
// slow down as the objectIndexer might be getting "too full". If Target Traverser is running
// slow causing objectIndexer map to get full, we induce wait to let Target Traverser catch up.

Reply:
But tomar if i don't call autopacer for target then all go routines will exit as there will be no work.